File: bitlap-core/src/main/kotlin/org/bitlap/core/sql/parser/ddl/SqlShowDatabases.kt
/* Copyright (c) 2023 bitlap.org */
package org.bitlap.core.sql.parser.ddl

import org.apache.calcite.DataContext
import org.apache.calcite.sql.SqlKind
import org.apache.calcite.sql.SqlSpecialOperator
import org.apache.calcite.sql.SqlWriter
import org.apache.calcite.sql.parser.SqlParserPos
import org.apache.calcite.sql.type.SqlTypeName
import org.bitlap.core.sql.parser.BitlapSqlDdlNode

/**
 * Desc:
 *    Parse tree for `SHOW (DATABASES | SCHEMAS)` statement.
 *
 * Mail: chk19940609@gmail.com
 * Created by IceMimosa
 * Date: 2021/8/25
 */
class SqlShowDatabases(pos: SqlParserPos) : BitlapSqlDdlNode(pos, OPERATOR, emptyList()) {

    companion object {
        val OPERATOR = SqlSpecialOperator("SHOW DATABASES", SqlKind.OTHER)
    }

    override fun unparse(writer: SqlWriter, leftPrec: Int, rightPrec: Int) {
        writer.keyword("SHOW DATABASES")
    }

    override val resultTypes: List<Pair<String, SqlTypeName>>
        get() = listOf(
            "database_name" to SqlTypeName.VARCHAR
        )

    override fun operator(context: DataContext): List<Array<Any?>> {
        return catalog.listDatabases().map { arrayOf(it.name) }
    }
}

File: bitlap-core/src/main/scala/org/bitlap/core/sql/parser/ddl/SqlShowDatabases.scala
/**
 * Copyright (C) 2023 bitlap.org .
 */
package org.bitlap.core.sql.parser.ddl

import org.bitlap.core.sql.parser.BitlapSqlDdlNode

import org.apache.calcite.DataContext
import org.apache.calcite.sql.`type`.SqlTypeName
import org.apache.calcite.sql.{ SqlIdentifier, SqlKind, SqlNode, SqlSpecialOperator, SqlWriter }
import org.apache.calcite.sql.parser.SqlParserPos

/** Parse tree for `SHOW (DATABASES | SCHEMAS)` statement.
 */
class SqlShowDatabases(override val _pos: SqlParserPos, val pattern: SqlIdentifier)
    extends BitlapSqlDdlNode(_pos, SqlShowDatabases.OPERATOR, List.empty[SqlNode]) {

  override def unparse(writer: SqlWriter, leftPrec: Int, rightPrec: Int): Unit = {
    writer.keyword("SHOW DATABASES")
  }

  override val resultTypes: List[(String, SqlTypeName)] = List(
    "database_name" -> SqlTypeName.VARCHAR
  )

  override def operator(context: DataContext): List[Array[Any]] = {
    catalog.listDatabases().map { it => Array(it.name) }
  }
}

object SqlShowDatabases {
  val OPERATOR: SqlSpecialOperator = SqlSpecialOperator("SHOW DATABASES", SqlKind.OTHER)
}
